Multi-Boot Fixes

DanTheMan827 opened the pull request:
Fixes are as follow:

**Exclude clovercon modules during mod_repair_modules**
This enables switching hsqs without breaking the controller hacks

**Separate game and clover directories based on the image type currently booted.**
This is needed because switching to a Japanese system and back caused a soft-brick in `ReedPlayer-Clover` until the `/var/lib/clover/` folder was deleted.

As a side-effect, this means that factory resets are also separate.

@ClusterM You may want to amend the save manager to look for saves in these directories as well so that users can manage them even if they're not booted into that image.

- `/var/lib/hakchi/foreign-clover/nes-usa/`
- `/var/lib/hakchi/foreign-clover/nes-jpn/`
- `/var/lib/hakchi/foreign-clover/snes-usa/`
- `/var/lib/hakchi/foreign-clover/snes-jpn/`

**Save a copy of the base hardware info to `/var/lib/hakchi/`, files include `boardtype`, `REGION`, and `hardware-id` (this is equal to `$sftype-$sfregion`)**
This information is needed from within the hsqs to properly configure clovercon, for example, booting the Famicom image on a SNES.

**Make hakchi2 use absolute game paths instead of `/usr/share/games`**
This makes it possible to sync a hsqs file as a "game" within hakchi and boot it.

@ClusterM when uninstalling hakchi, check for the presence of `/var/lib/foreign-clover/`, if it exists, ask the user if they want to erase the non-native saves.

[hakchi-2.21d-dantheman827-2c93f77.zip](https://github.com/ClusterM/hakchi2/files/1404545/hakchi-2.21d-dantheman827-2c93f77.zip)

DanTheMan827 committed:
Per system games directories

This script will move the games directory to a system-specific one if needed and do a bind mount to $rootfs/usr/share/games

This will keep SNES and NES games separate when switching between images and also prevent hakchi from accidentally deleting games for the other system.

DanTheMan827 committed:
Exclude clvcon.ko and clovercon.ko from mod_repair_modules

DanTheMan827 committed:
Save physical hardware info to reference when in a hsqs image

DanTheMan827 committed:
Per system clover directories

DanTheMan827 committed:
Reference the actual hardware boardtype for S78clvcon instead of the what squashfs says

madmonkey1907 commented:
>>Exclude clovercon modules during mod_repair_modules

controller won't work unless clovercon is installed. I'll fix this issue later.

DanTheMan827 commented:
Isn’t clovercon a separate hmod with its own rootfs?

Or is mod_repair_modules part of the hmod installation?

The only place I found mod_repair_modules was in load_firmware

DanTheMan827 commented:
@madmonkey1907 I just did a complete uninstall of hakchi and flashed back to stock kernel, I then flashed back to custom kernel and the gamepad hack was still functional in _both_ system images I tested (SNES and NES)

I don't prevent clovercon from installing, I prevent the rsync operation from copying `clvcon.ko` and `clovercon.ko` from squashfs to rootfs

madmonkey1907 commented:
yeah, but I don't have clovercon installed in the first place. open shell, type 

hakchi pack_uninstall clovercon

if you have snes, setup so nes image is loaded, after reboot, you will have no input.

DanTheMan827 commented:
ah, so it has to check if clovercon is installed before deciding to exclude it or not...

A simple if else should take care of that

DanTheMan827 commented:
@madmonkey1907 What about this? or would you do it another way?

```
mod_repair_modules(){
  if [ -f "$installpath/hmod/uninstall-clovercon" ]; then
    restore "/lib/modules/$(uname -r)/" "--exclude "*/clvcon.ko" --exclude "*/clovercon.ko""
  else
    restore "/lib/modules/$(uname -r)/"
  fi
}
```

madmonkey1907 commented:
I was thinking of setting up "/lib/modules/$(uname -r)/" as tmpfs in this case.

DanTheMan827 commented:
So, setup `/lib/modules/$(uname -r)/` as tmpfs, copy everything from squash modules to there, and have another preinit.d script for the clovercon mod to make sure it's always there?

Am I understanding that correctly?

Feel free to fork the branch and add commits

madmonkey1907 commented:
>So, setup /lib/modules/$(uname -r)/ as tmpfs, copy everything from squash modules to there

this part is correct
no modification to clovercon is required, except boardtype, but that can be done without temporary file, through cfg_* variables.

DanTheMan827 commented:
Keep in mind that the clovercon modules would also need to be copied from the mod to `/lib/modules/$(uname -r)/` if it's installed, that's where I was going with changes to the clovercon hmod

If you just copied the stuff from squash, the modded clovercon would be excluded

madmonkey1907 commented:
They don't need to be copied or overmounted. insmod is enough already. For snes module name need to be changed though.

DanTheMan827 committed:
Don't hardcode the paths into .desktop, it breaks on folders

Instead, for /bin/hsqs, expand /usr/share/games/ to the proper path in /var/lib/hakchi/games

DanTheMan827 committed:
Don't compress hsqs files and add it to the extension filter

DanTheMan827 committed:
Forgot to include two preinit.d scripts in the .csproj file

DanTheMan827 commented:
So, just add `mount -t tmpfs -o mode=0755,nosuid,nodev tmpfs "/lib/modules/$(uname -r)/"` into the `load_firmware` function and add a check so that [S78clvcon](https://github.com/ClusterM/hakchi2/blob/stable/mods/mod_hakchi/hakchi/rootfs/etc/init.d/S78clvcon) loads from rootfs first?

If modules is mounted as tmpfs before repairing, it wouldn't touch any possible hacked ones in rootfs

DanTheMan827 commented:
@madmonkey1907 Not sure if you saw my last comment.

Just wondering what your thoughts on that approach are.

kyleroberson commented:
Dan mind making a new build now that v2.21f is out?

DanTheMan827 commented:
@kyleroberson 
[hakchi2_2.21f_dantheman827_cad9a11.zip](https://github.com/DanTheMan827/hakchi2/files/1420471/hakchi2_2.21f_dantheman827_cad9a11.zip)

DanTheMan827 commented:
@ClusterM What do you think about me adding commit 979fd88 to this pull?

`pc000_clovercon` runs after `load_firmware`, and by extension `mod_repair_modules`, so it should fix hacked clovercon being overwritten without breaking the core clovercon if the hacked one is for some reason not installed.

DanTheMan827 commented:
@madmonkey1907 couple questions on this...

- Do you think I should change `p7075_profilebind` to store the foreign clover directories in `/var/lib/clover/foreign/` instead of `/var/lib/foreign-clover`? that way the "foreign" saves would be cleared with a factory reset on stock kernel. I don't think storing in `/var/lib/hakchi/` is a good idea for obvious reasons...
- Do you think this overall would be better as a hmod or included in core hakchi? I could probably do some overmount trickery to get a custom `/bin/hsqs` that expands the path and then run the original `/bin/hsqs` with the new path.

madmonkey1907 commented:
separate configs per image should be optional, not mandatory. and it's a bit too complex, especially if you consider that games subfolder might not be on nandc.
```
root@madmonkey:~# mount
rootfs on / type rootfs (rw)
devtmpfs on /dev type devtmpfs (rw,relatime,size=116860k,nr_inodes=29215,mode=755)
/dev/mapper/root-crypt on / type squashfs (ro,noatime)
tmpfs on /var type tmpfs (rw,nosuid,nodev,relatime,mode=755)
/dev/nandc on /var/lib type ext4 (rw,nosuid,nodev,noatime,stripe=32,data=ordered)
/dev/mapper/root-crypt on /var/lib/hakchi/squashfs type squashfs (ro,noatime)
/dev/nandc on /lib/modules type ext4 (rw,nosuid,nodev,noatime,stripe=32,data=ordered)
/dev/nandc on /bin type ext4 (rw,nosuid,nodev,noatime,stripe=32,data=ordered)
/dev/nandc on /etc type ext4 (rw,nosuid,nodev,noatime,stripe=32,data=ordered)
/dev/nandc on /usr/share/ui/snes-eur/resources/fonts/title.font type ext4 (rw,nosuid,nodev,noatime,stripe=32,data=ordered)
/dev/sda1 on /media type vfat (ro,relatime,fmask=0022,dmask=0022,codepage=cp437,iocharset=iso8859-1,shortname=mixed,errors=remount-ro)
/dev/sda1 on /usr/share/games type vfat (ro,relatime,fmask=0022,dmask=0022,codepage=cp437,iocharset=iso8859-1,shortname=mixed,errors=remount-ro)
/dev/nandc on /usr/bin/clover-canoe-shvc type ext4 (rw,nosuid,nodev,noatime,stripe=32,data=ordered)
proc on /proc type proc (rw,relatime)
none on /proc/bus/usb type usbfs (rw,relatime)
devpts on /dev/pts type devpts (rw,relatime,gid=5,mode=620,ptmxmode=000)
tmpfs on /dev/shm type tmpfs (rw,relatime,mode=777)
tmpfs on /tmp type tmpfs (rw,relatime)
tmpfs on /run type tmpfs (rw,nosuid,nodev,relatime,mode=755)
sysfs on /sys type sysfs (rw,relatime)
```

DanTheMan827 commented:
@madmonkey1907 Okay, now you've got me intrigued... `/dev/sda1`, is that a working external storage device?

As far as a separate config per image, I had a issue where booting into the Famicom image on a EUR Snes Classic corrupted the configuration for `ReedPlayer-Clover` resulting in just a black screen that persisted even if the stock kernel was flashed.

It turned out to be a corrupted profile that was probably caused by switching between images.

As long as the image had English, sharing a profile was fine, but as soon as I booted the Japanese image,  it broke... I never quite figured out exactly what did it, but that's what happened.

madmonkey1907 commented:
working mass-storage device. if you have usb-otg hub, can send you a test version.

I didn't experience what you describe with famicom image. do you mean super famicom? there is an ongoing problem with retroarch module stemming from incorrect handling of filenames, maybe it's because of that.

DanTheMan827 commented:
@madmonkey1907 okay, so I can't recreate that problem either... huh... well, at the least, isolating the profiles prevents the language select from coming up on every switch...

I do not have a USB OTG hub... even if I did though, am I correct in assuming that clovershell can't work with USB storage attached? (I have no UART connection to the SNES)

If USB storage can work **and** clovershell, that's pretty amazing... although SD would be a nice alternative to keep things nice and tidy...

madmonkey1907 commented:
Yeah, clovershell won't work. It never worked on Linux though, so not a big deal?

DanTheMan827 commented:
@madmonkey1907 Well, I use Windows and Hakchi2 primarily... I haven't even used your hakchi app to be honest...

Did USB mass storage require a kernel mod, or just a USB hub?

I suppose enabling the SD card would be a bit more difficult without some kind of shell into the system, huh...

Is there any alternative method for a shell other than UART1 or clovershell?

madmonkey1907 commented:
I jest. Instead of hauling your (s)nes classic from tv to pc and back, you could just haul usb stick or sd card. and clovershell will work, if alternate kernel is booted over usb. kexec might work too.

There are no sd or usb-storage drivers in stock kernel. end users though, they will only need otg hub.

well, there should be UART2 at the very least. not really a problem if you're going to solder anyway.

DanTheMan827 commented:
I decided to move this functionality into a hmod... Hakchi2 still does need a tweak in the file chooser though, hsqs isn't a valid option and it compresses them if you force select one.

https://www.reddit.com/r/miniSNESmods/comments/7agl3m/multiboot_fixes_hmod_dual_boot_with_working/
